controller/pods.go

```diff
@@ -51,6 +53,113 @@ const (
 	retryDelay = 1 * time.Second
 )
 
+type OpType int
+
+const (
+    Add OpType = iota
+    Delete
+)
+
+// Event struct
+type Event struct {
+	opType OpType
+	cniParams *cni.Parameters
+}
+
+// eventQueue to publish events to
+type EventQueue struct {
+	q *list.List
+	m map[string]*list.Element  // ref in the eventQueue list
+	lock sync.Mutex
+	cond *sync.Cond
+}
+
+func newQueue() *EventQueue {
+	eq := &EventQueue{m: make(map[string]*list.Element), q: list.New()}
+	eq.q.Init()
+	eq.cond = sync.NewCond(&eq.lock) 
+	return eq
+}
+
+func (eq *EventQueue) Enqueue(opType OpType, cniParams *cni.Parameters) {
+	eq.cond.L.Lock()
+	defer eq.cond.L.Unlock()
+	
+	key := eq.getKey(cniParams)
+	glog.V(5).Infof("Enqeuing using key:%s", key)
+	if e, ok := eq.m[key]; ok {
+		ev := e.Value.(Event)
+		if ev.opType != opType {
+			glog.Infof("Cancelling events from queue - events cancels each others:", ev)
+			eq.q.Remove(e)
+			delete(eq.m, key)
+			return
+		}
+    }
+	
+	ev := Event{opType, cniParams}
+	e := eq.q.PushBack(ev)
+	glog.V(5).Infof("Enqueue new event:", ev)
+	eq.m[key] = e
+	eq.cond.Signal()
+}
+
+func (eq *EventQueue) Dequeue() *Event {
```

Comment:
Since the lock strategy is different for Dequeue and Enqueue, would add a comment that lock must be held when calling the Dequeue method